[김재환] week4

BBoBBoM opened the pull request:
## 요구사항

### 기본

- [x] [기본]이메일 input에서 focus out 할 때, 값이 없을 경우 input에 빨강색 테두리와 아래에 “이메일을 입력해주세요.” 빨강색 에러 메세지가 보이나요?
 
- [x] [기본]이메일 input에서 focus out 할 때, 이메일 형식에 맞지 않는 값이 있는 경우 input에 빨강색 테두리와 아래에 “올바른 이메일 주소가 아닙니다.” 빨강색 에러 메세지가 보이나요?
 
- [x] [기본]이메일 input에서 focus out 일 때, input 값이 test@codeit.com 일 경우 input에 빨강색 테두리와 아래에 “이미 사용 중인 이메일입니다.” 빨강색 에러 메세지가 보이나요?
 
- [x] [기본]비밀번호 input에서 focus out 할 때, 값이 8자 미만으로 있거나 문자열만 있거나 숫자만 있는 경우, input에 빨강색 테두리와 아래에 “비밀번호는 영문, 숫자 조합 8자 이상 입력해 주세요.” 빨강색 에러 메세지가 보이나요?
 
- [x] [기본]비밀번호 input과 비밀번호 확인 input의 값이 다른 경우, 비밀번호 확인 input에 빨강색 테두리와 아래에 “비밀번호가 일치하지 않아요.” 빨강색 에러 메세지가 보이나요?
 
- [x] [기본]회원가입을 실행할 경우, 문제가 있는 경우 문제가 있는 input에 빨강색 테두리와 에러 메세지가 보이나요?
 
- [x] [기본]이외의 유효한 회원가입 시도의 경우, “/folder”로 이동하나요?
 
- [x] [기본]이메일: test@codeit.com, 비밀번호: codeit101 으로 로그인 시도할 경우, “/folder” 페이지로 이동하나요?
 
- [x] [기본]비밀번호 input에서 focus out 할 때, 값이 없을 경우 아래에 “비밀번호를 입력해주세요.” 에러 메세지가 보이나요?
 
- [x] [기본]이외의 로그인 시도의 경우 이메일, 비밀번호 input에 빨강색 테두리와 각각의 input아래에 “이메일을 확인해주세요.”, “비밀번호를 확인해주세요.” 빨강색 에러 메세지가 보이나요?

- [x] [기본]회원가입 버튼 클릭 또는 Enter키 입력으로 회원가입 되나요?

- [x] [기본]이메일, 비밀번호 input에 에러 관련 디자인을 Components 영역의 에러 케이스로 적용했나요?

### 심화

- [x] [심화]눈 모양 아이콘 클릭시 비밀번호의 문자열이 보이기도 하고, 가려지기도 하나요?
- [x] [심화]비밀번호의 문자열이 가려질 때는 눈 모양 아이콘에는 사선이 그어져있고, 비밀번호의 문자열이 보일 때는 사선이 없는 눈 모양 아이콘이 보이나요?
- [x] [심화]로그인, 회원가입 페이지에 공통적으로 사용하는 로직이 있다면, 반복하지 않고 공통된 로직을 모듈로 분리해 사용했나요?

## 주요 변경사항
### 네틀리파이 배포 주소 : [https://week4-codit-jaehwan.netlify.app/](url)
-클래스 추가, 제거에 관한 함수
-input태그 placeholder 추가 사항
-emailInput 관한 css스타일 id pattern check 및 오류 메시지
-passwordInput 관한 css스타일 id pattern check 및 오류 메시지
-비밀번호 입력란의 눈모양 클릭시 이미지 변경 및 type 변경으로 글자표시 가림을 확인할수 있는 코드
-로그인,회원가입에 관련한 요구사항 구현 및 페이지 이동 잘못입력시 각 input 태그 css변경 및 오류메시지
-이메일 중복체크
-회원가입 버튼에 관련한 요구사항 구현 및 페이지 이동 잘못입력시 각 input 태그 css변경 및 오류메시지
-비밀번호 형식체크 및 비밀번호 비밀번호 확인 input 비교해서 일치 불일치 확인
-passwordInput 관한 css스타일 id pattern check 및 오류 메시지
- 모듈화 작업 구현

## 스크린샷
![1](https://github.com/codeit-bootcamp-frontend/2-Weekly-Mission/assets/131963909/ec6226b4-5da2-4e5d-9b3c-6525828c6c7d)
1.로그인 페이지 email과 비밀번호 미입력시 오류메시지 알림 

![2](https://github.com/codeit-bootcamp-frontend/2-Weekly-Mission/assets/131963909/93579da7-f3a2-4ebd-ae74-c3333f48bc77)
2.이메일 형식 오류와 눈모양 이미지 클릭시 이미지 변화와 textshow 확인

![3](https://github.com/codeit-bootcamp-frontend/2-Weekly-Mission/assets/131963909/e1a8211c-8fe9-4ccc-a27b-b1ef06de1bbc)
3.회원가입시 각 input태그 텍스트 미입력시 발생하는 오류 

![4](https://github.com/codeit-bootcamp-frontend/2-Weekly-Mission/assets/131963909/976b16cc-5a46-4009-8b59-5377a2901452)
4. 이메일 형식, 비밀번호 형식, 비밀번호와 비밀번호 확인간의 비밀번호 일치여부, textshow버튼 활성화 확인

![5](https://github.com/codeit-bootcamp-frontend/2-Weekly-Mission/assets/131963909/f219408e-a95b-425a-bb59-a903b9f8ff28)
5. 로그인버튼 클릭과 enter submit시 기본동작을 막고 오류메시지 안내

![6](https://github.com/codeit-bootcamp-frontend/2-Weekly-Mission/assets/131963909/50f45eca-f465-4c01-acf3-4c02dff0e0d0)
6. 기본 부여된 이메일 비밀번호 제외 다른 형식에 맞는 이메일 비밀번호를 입력했을때 오류가 나지 않는 확인사항

![7](https://github.com/codeit-bootcamp-frontend/2-Weekly-Mission/assets/131963909/e4db8d5a-8ee0-473b-9f59-8b722541f4e3)
7. 6번을 submit 했을때 넘어가는 folder.html



## 멘토에게

- 아직 모듈로 분리작업이 많이 어렵고 감이 잡히질 않습니다.(예를 들면 어디까지 export 하고 import 해온 js파일에서 변수를 활용하는 부분에서 혼란이 많아 흐름 유지하는 것이 어렵긴했지만 어찌저찌 구현이 된것 같습니다.
- 학습 부족으로 모던 자바스크립트 문법의 적용이 어려워 대부분 기본형식으로 작업했습니다.
- 코드 흐름을 유지하는 부분에서 어려움을 느꼈습니다. (세분화하고 해봐도 사고방식이 수준에 못 미쳐 흐름을 중간중간 잃고 자꾸 헤매는 시간이 많았습니다
- 요구사항중  "이외의 로그인 시도의 경우 이메일, 비밀번호 input에 빨강색 테두리와 각각의 input아래에 “이메일을 확인해주세요.”, “비밀번호를 확인해주세요.” 빨강색 에러 메세지가 보이나요?" 
 1. 이부분의 해석이 모호해 로그인 버튼을 누르면  모든 경우를 생각해서 오류 메시지를 보여줘야 하나에 대한 의문과
 2. 아이디 체크나 비밀번호 형식체크 확인과의 일치 작업에서의 오류메시지 focusout 시 동작하도록 했는데 작업을 해야하나에 대한 고민이었습니다.
 3.  우선 비밀번호를 확인해주세요에대한 text변경으로 동작은 하게 해놓았으나 이것이 제대로 구현이된 부분인지는 여전히 의문입니다.
- 요구사항 중 회원가입 버튼 클릭 또는 Enter키 입력으로 회원가입 되나요? 에 관한 질문입니다. form에서는 enter키를 누르면 자동 submit이 동작하는 부분으로 알고 있고 실제로도 Enter 키 등록을 안했는데 동작을 확인했습니다. 위의 요구사항으로 보아 새로 만들어야하는 의구심이 들었습니다 혹시 몰라 맨 아래 기본형식은 구현해놓고 주석처리 해놓았으나 ... 저의 짧은 소견으로는 의도 파악이 잘 안됩니다.
- 제가 정리할수 있는 한 최대한 깔끔하고 중복없이 버려지는 코드 없이 하려고 하였으나 많이 부족한거 같습니다.
- 이번주 과제는 어렵기는 했지만 무척 재미있었고 배운것들 실제 적용 함으로써 조금의 성장도 느꼈습니다.! 
- 모듈화 이해가 어렵고 실제 적용하려니 그 방법의 어려움이 있었는데 재미있게 했던것 같습니다 멘토님께서 생각하시는 방향이나 구현방법에 다른방법이나 효율적인 부분이 있다면 알려주세요 지금은 강의에서 배운내용만으로 어렴풋이 구현을 했습니다.

![5](https://github.com/codeit-bootcamp-frontend/2-Weekly-Mission/assets/131963909/f219408e-a95b-425a-bb59-a903b9f8ff28)에서 회원가입 버튼을 누르면 조건에 부합하지 않는 부분 테두리 표시와 오류 메시지 안내가 필요한데 이부분이 미흡한것 같습니다.
SignupButtonEvent.js 파일 입니다 저는 아이디의 값이 없다면 아이디 이력에 관한 오류메시지, 중복에 관한 메시지를 별개로 표현하고 싶었는데 이부분에 대한 사고가 미치지 못해 해메고 있어요

BBoBBoM committed:
4주차 위클리 미션입니다.

지난번 피드백도 있고 이번에는 제공해주는 기본 코드에 이번주 과제인
signin에 관한 요구사항을 구현했습니다. script폴더 signin.js 에서
작업했습니다.

BBoBBoM committed:
회원 가입페이지 일부 구현

로그인 버튼 눌렀을때의 조건에 부합하지 않는 input표시 작업 미구현
모듈 분리 미구현

BBoBBoM committed:
일부 조그마한 오류부분 수정

BBoBBoM committed:
미동작하는 부분 일부 수정

BBoBBoM committed:
모듈 구현

제대로 한것인가에 대한 의문 .... signup에서의 회원가입 버튼 오류 메지시
보여주는 부분 미흡 추후 작업예정

Aurumsapiens commented:
- 아직 모듈로 분리작업이 많이 어렵고 감이 잡히질 않습니다.(예를 들면 어디까지 export 하고 import 해온 js파일에서 변수를 활용하는 부분에서 혼란이 많아 흐름 유지하는 것이 어렵긴했지만 어찌저찌 구현이 된것 같습니다.
-> 고생 많으셨습니다! 말씀해주신대로 고민의 흔적이 남아 있는 듯 합니다. 지금 당장은 좋은 결과물보다 좋은 고민이 성장에 더 도움이 되지않을까합니다~!

- 학습 부족으로 모던 자바스크립트 문법의 적용이 어려워 대부분 기본형식으로 작업했습니다.
코드 흐름을 유지하는 부분에서 어려움을 느꼈습니다. (세분화하고 해봐도 사고방식이 수준에 못 미쳐 흐름을 중간중간 잃고 자꾸 헤매는 시간이 많았습니다
-> 구현을 하는 것이 일단은 더 중요하다고 생각합니다~ 작업을 잘 진행해주신 것으로 보입니다~!

- 요구사항중 "이외의 로그인 시도의 경우 이메일, 비밀번호 input에 빨강색 테두리와 각각의 input아래에 “이메일을 확인해주세요.”, “비밀번호를 확인해주세요.” 빨강색 에러 메세지가 보이나요?"
1. 이부분의 해석이 모호해 로그인 버튼을 누르면 모든 경우를 생각해서 오류 메시지를 보여줘야 하나에 대한 의문과
-> 음.. 제가 이해한 걸로는 이메일 형식으로 입력을 하고 비밀번호를 입력한 상황에서 로그인 버튼을 눌렀을 때, /folder 경로로 이동할 수 없는 계정이라면 에러메세지를 출력하는 것으로 이해를 했습니다.
2. 아이디 체크나 비밀번호 형식체크 확인과의 일치 작업에서의 오류메시지 focusout 시 동작하도록 했는데 작업을 해야하나에 대한 고민이었습니다.
-> 음.. 이거는 제가 이해를 잘 못해서 멘토링 시간에 한번 이야기를 해보면 좋을 듯 합니다.
3. 우선 비밀번호를 확인해주세요에대한 text변경으로 동작은 하게 해놓았으나 이것이 제대로 구현이된 부분인지는 여전히 의문입니다.
-> 잘 구현된 것으로 보이는데 이 또한 멘토링 시간에 추가적으로 이야기 해보겠습니다.

- 요구사항 중 회원가입 버튼 클릭 또는 Enter키 입력으로 회원가입 되나요? 에 관한 질문입니다. form에서는 enter키를 누르면 자동 submit이 동작하는 부분으로 알고 있고 실제로도 Enter 키 등록을 안했는데 동작을 확인했습니다. 위의 요구사항으로 보아 새로 만들어야하는 의구심이 들었습니다 혹시 몰라 맨 아래 기본형식은 구현해놓고 주석처리 해놓았으나 ... 저의 짧은 소견으로는 의도 파악이 잘 안됩니다.
-> 4주차가 아닌 5주차 요구사항을 말씀하시는 거였군요! 이 또한 한번 멘토링 시간에 이야기해보면 좋을 듯 합니다.

- 제가 정리할수 있는 한 최대한 깔끔하고 중복없이 버려지는 코드 없이 하려고 하였으나 많이 부족한거 같습니다.
-> 그렇게 작업하시려고 노력하신 점이 높이 살만하다고 생각합니다 :)

- 이번주 과제는 어렵기는 했지만 무척 재미있었고 배운것들 실제 적용 함으로써 조금의 성장도 느꼈습니다.!
-> Goooood!!!

- 모듈화 이해가 어렵고 실제 적용하려니 그 방법의 어려움이 있었는데 재미있게 했던것 같습니다 멘토님께서 생각하시는 방향이나 구현방법에 다른방법이나 효율적인 부분이 있다면 알려주세요 지금은 강의에서 배운내용만으로 어렴풋이 구현을 했습니다.
-> 저는 이번에 작업해주신 내용이 정말 좋다고 생각합니다. 좋은 코드인지를 떠나서 모듈화를 위해 고민하고 직접 코드를 짜신 모습이 좋다고 생각합니다!

Aurumsapiens commented on signin/signin.html:
```diff
@@ -1,66 +1,71 @@
 <!DOCTYPE html>
```
현재 파일명이 signin.html이지만 sigin/index.html 로 수정해보는건 어떨까요~? 그리고 이렇게 수정이 되면 어떻게 변할까요?!

BBoBBoM replied:
최상위의 index에서 signin 으로 안넘어가고 .... 흠....  동작부분에서는 특이사항없는것같습니다만.... 멘토님이 어떤 의도로 요청한것인지 이해를 못했습니다!

Aurumsapiens commented on script/signin.js:
```diff
@@ -0,0 +1,26 @@
+
+
+//input태그 placeholder 추가 사항
+emailInput.setAttribute("placeholder", "E-mail을 입력해주세요");
+passwordInput.setAttribute("placeholder", "비밀번호를 입력해주세요");
```
혹시 html에서 attribute로 설정하지않고 js로 설정하신 이유가 있으실까요~?

BBoBBoM replied:
이유는 없고 css나 html 에서 구현해도 되지만 css속성도 익히고 작은부분이라도 실습하고자 일부러 js에서 속성을 부여했습니다.

Aurumsapiens commented on function/AddRemoveFunction.js:
```diff
@@ -0,0 +1,49 @@
+  passwordErrorCheck,
+} from "../tag.js";
+
+function emailAddClass() {
```
함수명은 가능하면 동사가 먼저 있으면 좋을 듯 합니다~
addClassToEmailElement 혹은 addClassEmail 등등의 네이밍을 사용하는것을 더 추천드립니다!
다른 함수들도 마찬가지일 듯 합니다~!

https://sooftware.io/coding-habit/

BBoBBoM replied:
넵 알려주신 사이트 참고해서 함수명 변수명 짓는 연습도 해보겠습니다

Aurumsapiens commented on tag.js:
```diff
@@ -0,0 +1,27 @@
+const eyeImg = document.querySelector("#eye-img");
+const signInput2 = document.querySelector("#password-input-check");
+const eyeImg2 = document.querySelector("#eye-img2");
+let userInfo = [{ email: "test@codeit.com", password: "codeit101" }];
```
여기 함수는 기 정의된 엘리먼트들을 모아놓은 형태로 보입니다~
반면에 userInfo는 기존에 등록된 user 정보인 듯 하여 성향이 서로 다른듯합니다.
registeredUsers 와 같은 파일과 변수명으로 만들어 놓고 거기에 정리를 하는 것이 어떨까요~?

BBoBBoM replied:
말씀해주신 부분은 후의 DB와 연결해서 그 값으로 비교하는 부분 같은데 아직 어디에 어떤 부분을 넣고 분리해야하는지 어렵습니다 
성향이 다르다는말 이해했습니다 수정해보겠습니다

Aurumsapiens commented on function/eyeImgClickEvent.js:
```diff
@@ -0,0 +1,14 @@
+/*눈모양 아이콘 클릭시 사선이미지와 사선없는 이미지 교체와 그에 따른 input type 변경에 대한 함수*/
```
이 파일만 소문자로 시작을 하는 것 같은데 통일을 시켜주는 것이 좋을 듯 합니다. 함수명과 함수 파일들은 소문자로 시작하면서 동사로 시작하는 걸로 수정해보는 것이 어떨까요~?

BBoBBoM replied:
넵 알겠습니다 통일성 일관성 새겨듣겠습니다

Aurumsapiens commented on function/SignupButtonEvent.js:
```diff
@@ -0,0 +1,31 @@
+  if (userConfirmCheck && passwordConfirmCheck && passwordMachingCheck) {
+    window.location.href = "/folder/folder.html";
+  } else {
+    emailInputFocusoutEvent;
+    passwordInputFocusoutEvent;
+    userDuplicateCheck;
```
이거는 함수 실행을 해주어야하는 부분일까요~?

BBoBBoM replied:
제가 사실 이부분에서 고민이 있었는데 회원가입 버튼을 누르면 이메일 비밀번호 비밀번호 확인 부분중 오류가 있는 부분을 빨간테두리로 css부여하고 그아래 어떠한 유형의 오류인지 안내문구를 내고 싶었습니다. 
함수를 한번더 호출해서 각 인풋태그에 오류부분을 체크한다는 의미에서 함수 호출을 넣어봤습니다만 사실 이부분은 구현을 못했습니다

Aurumsapiens commented on function/PasswordCreateRule.js:
```diff
@@ -0,0 +1,24 @@
+    return true;
+  }
+}
+export default passwordInputFocusoutEvent;
```
어떤 함수 파일에서는 export default를 하거나 어떤 함수 파일에서는 export default를 하지 않는 것으로 보입니다~ 이것도 통일이 되면 좋을 듯 합니다!

BBoBBoM replied:
넵 알겠습니다 약간은 변명을 하자면 1개도 {} 방법과 default 방법이 있는 것을 보고 실습차 default는 {}방법과 다르게 import도 약간 다른것을 확인하고 익히고자 해서 통일성을 해치긴하지만 시도해봤습니다.

BBoBBoM committed:
registerUser.js 파일 추가

registerUser.js 파일의 userinfo가 다른 태그를 다루는 변수와 성향이
다르다는 멘토님의 의견에 따라 파일을 새로 생성
userinfo를 import 해서 사용하는 파일에서의 from위치변경작업업

BBoBBoM committed:
registerUser.js 파일 추가

registerUser.js 파일의 userinfo가 다른 태그를 다루는 변수와 성향이
다르다는 멘토님의 의견에 따라 파일을 새로 생성
userinfo를 import 해서 사용하는 파일에서의 from위치변경작업업

BBoBBoM committed:
Merge branch 'part1-김재환-week4' of https://github.com/BBoBBoM/2-Weekly-Mission into part1-김재환-week4
